[GLUTEN-10874][INFRA] Maturity RE50: In documentation, add a section for the release process

zhztheplayer opened the pull request:
#10874

The PR adds a documentation to demonstrate how to generate a complete release for a specific Gluten version locally.

The steps should be repeatable as long as the environment  meets the prerequisites.

PHILO-HE reviewed (commented):
@zhztheplayer, thanks for your work. Please check if my comments make sense.

PHILO-HE commented on dev/release/package-release.sh:
```diff
@@ -15,7 +15,7 @@
 # limitations under the License.
 
-# Note: Manually create $GLUTEN_HOME/release/ and place the release JARs inside.
+# Note: The script looks up the release JARs inside $GLUTEN_HOME/package/target/.
```
@zhztheplayer, thank you for the reply. I am not sure if users really need to create an official-like release by themselves. In official release, we build Gluten for all supported Spark versions with all feature modules enabled, even though generally some of them are not enabled together by users (e.g., Celeborn and Uniffle). However, most Gluten users only focus on one specific Spark version, and some of the the feature modules, which means they just need a clear build guide, not a release guide. Users like cloud service providers may require building releases for multi-Spark-version support. But they likely maintain their own processes and scripts to customize releases for forked repositories, which I assume differs significantly from our official release process.

If my understanding is correct, maybe we can focus only on the official release process, primarily for release managers' reference?

zhztheplayer replied:
Thanks for the insights. Now I think I am convinced we should go with a release manager guide first. 

RE50 itself was seen a bit ambiguous and you are right it might better to read it as "creating new releases" rather than "reproducing old releases".

PHILO-HE commented on dev/release/package-release.sh:
@zhztheplayer, could you please include the following changes?  I have applied these changes when publishing the recent 1.5.0 RC1. It's for meeting the release requirement (The approved release candidate will be directly converted to a formal release. So "rc*" shouldn't be used in naming packaged folders).

```diff
diff --git a/dev/release/package-release.sh b/dev/release/package-release.sh
index ec18101ae..d60a14e78 100755
--- a/dev/release/package-release.sh
+++ b/dev/release/package-release.sh
@@ -53,11 +53,14 @@ for v in $SPARK_VERSIONS; do
 done

 SRC_ZIP="${TAG}.zip"
-SRC_DIR="incubator-gluten-${TAG_VERSION}"
+SRC_DIR="incubator-gluten-${RELEASE_VERSION}"

 echo "Packaging source code..."
 wget https://github.com/apache/incubator-gluten/archive/refs/tags/${SRC_ZIP}
 unzip -q ${SRC_ZIP}
+
+# Rename folder to remove "rc*" for formal release.
+mv incubator-gluten-${TAG_VERSION} ${SRC_DIR}
 tar -czf apache-gluten-${RELEASE_VERSION}-incubating-src.tar.gz ${SRC_DIR}
 rm -r ${SRC_ZIP} ${SRC_DIR}
```

PHILO-HE commented on docs/developers/HowToRelease.md:
Sounds good to me. Thanks.

PHILO-HE commented on docs/developers/HowToRelease.md:
```diff
@@ -0,0 +1,76 @@
+layout: page
+title: How to Release
+nav_order: 17
+parent: Getting-Started
```
Please update accordingly. Thanks.

zhztheplayer replied:
> It seems there are quite some issues with the current nav orders.

Unfortunately, after a quick check, I think I need more time for doing this.

I've update the PR to make the new page correctly show up, so we can merge this first.

zhztheplayer commented:
Here is the update of the PR:

1. I've left `package-release.sh` nearly intact and changed the `HowToRelease.md` to do an extra copy from `package/target` to `release`
2. I've moved the release publication guid from `incubator-gluten-site`'s 1.3 branch to this documentation

@PHILO-HE And feel free to help update `HowToRelease.md` to use the CI job to obtain release binaries afterwards. This PR should close RE50 then.
